src/utils/EntryModel.hxx

```diff
@@ -41,389 +41,465 @@
+    void init(unsigned max_size, int base)
     {
-        HASSERT(digits <= N);
-        digits_ = digits;
+        maxSize_ = max_size;
         clear();
+        set_base(base);
+        set_boundaries();
+    }
 
-        string str;
-        switch (base)
+    /// Initialize with a value.
+    /// @param max_size max number of digits in the base type
+    /// @param base base type, 10 or 16
+    /// @param value value to initialize with
+    void init(unsigned max_size, int base, T value)
+    {
+        init(max_size, base);
+        value_ = value;
+        isAtInitialValue_ = true;
+        empty_ = false;
+    }
+
+    /// Append a value to the "back".
+    /// @param val value to append, base 10: 0 - 9, base 16: 0x0 - 0xF
+    void push_back(uint8_t val)
+    {
+        HASSERT(val < base_);
+        if (size_ >= maxSize_)
         {
-            default:
-                HASSERT(0);
-            case 10:
-                if (std::is_signed<T>::value)
-                {
-                    str = int64_to_string(value, digits);
-                }
-                else
-                {
-                    str = uint64_to_string(value, digits);
-                }
-                break;
-            case 16:
-                if (std::is_signed<T>::value)
-                {
-                    str = int64_to_string_hex(value, digits);
-                }
-                else
-                {
-                    str = uint64_to_string_hex(value, digits);
-                }
-                if (transform_)
-                {
-                    /* equires all characters in upper case */
-                    transform(str.begin(), str.end(), str.begin(), toupper);
-                }
-                break;
+            // clear entry, return without appending the character
+            clear();
+            return;
         }
-        str_populate(data_, str.c_str());
-        hasInitial_ = true;
+        if (isAtInitialValue_)
+        {
+            // clear entry before inserting character
+            clear();
+        }
+        value_ *= base_;
+        if (value_ < 0)
+        {
+            value_ -= val;
+        }
+        else
+        {
+            value_ += val;
+        }
+        if (value_ == 0 && !empty_)
+        {
+            if (numLeadingZeros_ < 31)
+            {
+                ++numLeadingZeros_;
+            }
+        }
+        empty_ = false;
+        ++size_;
+        clamp();
     }
 
-    /** Clear the entry string.
-     * @param data data to fill in the buffer with
-     */
-    void clear(const char *data = nullptr)
+    /// Append a character to the "back".
+    /// @param c character to append, base 10: 0 - 9, base 16: 0 - F
+    void push_back_char(char c)
     {
-        memset(data_, ' ', digits_);
-        if (data)
+        switch (base_)
         {
-            memcpy(data_, data, strlen(data));
+            case 10:
+                HASSERT(c >= '0' && c <= '9');
```

Comment:
it is a bit scary to crash like this, maybe we should just ignore invalid input (do nothing)

Reply:
Actually, this is what I prefer. We should get here based on incorrect runtime code in the same executable only.